xds/src/main/java/io/grpc/xds/GrpcAuthorizationEngine.java

```diff
@@ -0,0 +1,369 @@
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return delegate.matches(args.getDestinationIp());
+    }
+  }
+
+  static final class SourceIpMatcher extends Matcher {
+    private final CidrMatcher delegate;
+
+    SourceIpMatcher(CidrMatcher delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return delegate.matches(args.getSourceIp());
+    }
+  }
+
+  static final class PathMatcher extends Matcher {
+    private final StringMatcher delegate;
+
+    PathMatcher(StringMatcher delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return delegate.matches(args.getPath());
+    }
+  }
+
+  static final class HeaderMatcher extends Matcher {
+    private final Matchers.HeaderMatcher delegate;
+
+    HeaderMatcher(Matchers.HeaderMatcher delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return delegate.matches(args.getHeader());
+    }
+  }
+
+  static final class DestinationPortMatcher extends Matcher {
+    private final int port;
+
+    DestinationPortMatcher(int port) {
+      this.port = port;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return port == args.getDestinationPort();
+    }
+  }
+
+  private static final class EvaluateArgs {
+    private final Metadata metadata;
+    private final ServerCall<?,?> serverCall;
+    private static final int URI_SAN = 6;
+    private static final int DNS_SAN = 2;
+    // https://github.com/envoyproxy/envoy/blob/63619d578e1abe0c1725ea28ba02f361466662e1/api/envoy/config/rbac/v3/rbac.proto#L238-L240
+    // Set the principal name for SAN types in this order, higher priority on the left. If none of
+    // them present, subject name is the fallback.
+    private static final ImmutableList<Integer> SAN_TYPES = ImmutableList.of(URI_SAN, DNS_SAN);
+
+    private EvaluateArgs(Metadata metadata, ServerCall<?,?> serverCall) {
+      this.metadata = metadata;
+      this.serverCall = serverCall;
+    }
+
+    private String getPath() {
+      return serverCall.getMethodDescriptor().getFullMethodName();
+    }
+
+    @Nullable
+    private String getPrincipalName() {
+      SSLSession sslSession = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_SSL_SESSION);
+      try {
+        Certificate[] certs = sslSession == null ? null : sslSession.getPeerCertificates();
+        if (certs == null || certs.length < 1) {
+          return null;
+        }
+        X509Certificate cert = (X509Certificate)certs[0];
+        if (cert == null || cert.getSubjectDN() == null) {
+          return null;
+        }
+        String principalName = cert.getSubjectDN().getName();
+        Collection<List<?>> names = cert.getSubjectAlternativeNames();
+        int chosenIndex = Integer.MAX_VALUE;
+        if (names != null) {
+          for (List<?> name : names) {
```

Comment:
This isn't right. If there is a URI SAN, then you should not do any checks for a DNS SAN.

Reply:
hm, I checked for both in one run but I will discard DNS and replace it with URI if I find URI. Alternatively yes I can do two passes.

Reply:
Oh, that's what the `chosenIndex` is about. Sorry, it was the "wrong shape" so I didn't pay that much attention.

And then you short-circuit if you find a URI early... Except, that's not the new behavior I described. The new behavior I described checks _all_ URI SANs to see if they match. You have to check for matches against multiple values. `getPrincipalName()` isn't the right API; it needs to be `Collection<String> getPrincipalNames()`. The Envoy code has has three methods, one for each type. Or just return the cert and let the matcher find what it is interested in.

Even if that the above wasn't a problem, I don't really see why this index stuff is necessary. Having two for loops or a nested loop, one for each type, is no slower nor complicated; `indexOf` is already a for loop. The most obvious implementation looks quite fair, although I admit it doesn't log and it could get out of hand if you needed a bunch of instanceof checks or other such details in each loop:

```java
// Remember, this is broken because we need to check all entries of a type
if (names != null) {
  for (List<?> name : names) {
    if (URI_TYPE.equals(name.get(0))) {
      return (String) name.get(1);
    }
  }
  for (List<?> name : names) {
    if (DNS_TYPE.equals(name.get(0))) {
      return (String) name.get(1);
    }
  }
}
return cert.getSubjectDN().getName();
```

And your code was going a bit more generic, this is the structure I expected when I saw `SAN_TYPES` defined (which is still pretty legible):
```java
// Remember, this is broken because we need to check all entries of a type
if (names != null) {
  for (Integer type : SAN_TYPES) { // All types assumed to have String values
    for (List<?> name : names) {
      if (type.equals(name.get(0))) {
        return (String) name.get(1);
      }
    }
  }
}
return cert.getSubjectDN().getName();
```

But that is neither here nor there.

Reply:
> Except, that's not the new behavior I described. The new behavior I described checks all URI SANs to see if they match.

`s/see if they match/see if any match` ?